devicemodel/hw/pci/virtio/virtio_ipu.c

```diff
@@ -0,0 +1,391 @@
+/*
+ * Copyright (C) 2018 Intel Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+/*
+ * virtio ipu
+ * IPU mediator DM
+ */
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <assert.h>
+#include <pthread.h>
+
+#include "dm.h"
+#include "pci_core.h"
+#include "virtio.h"
+#include "virtio_kernel.h"
+#include "vmmapi.h"
+
+/*
+ * Size of queue was chosen experimentally in a way
+ * that it allows to do IPU ISYS MIPI camera capture without
+ * any delay for interrupt/msg send, this value should
+ * be tuned.
+ */
+#define VIRTIO_IPU_RINGSZ 1024
+
+/*
+ * IPU mediation makes use of 2 VQs.
+ * VQ0 for buffer mgmnt & 1 time configuration;
+ * VQ1 for interrupt/msg exchange
+ */
+#define VIRTIO_IPU_VQ_NUM 2
+
+#define IPU_VBS_DEV_PATH "/dev/vbs_ipu"
+
+static int ipu_log_level = 0;
+#define TAG "virtio_ipu: "
+#define LERR 1
+#define LWRN 2
+#define LDBG 3
+#define UPRINTF(lvl, fmt, args...) \
+	do { if (lvl <= ipu_log_level) printf(TAG fmt, ##args); } while (0)
```

Comment:
The ipu_log_level is default 0, that is mean no any logs will be print, is it your expected ?
BTW, The UPRINTF is mean "USB PRINTF", you needn't to copy its name..  Maybe IPRINTF is more appropriate for IPU...

As I mentioned in last time, the ipu is not necessary to add the print level, the usb added it due to it included too much debug info.. How about:

static int ipu_log_level;
#define TAG "virtio_ipu: "
#define DPRINTF(fmt, args...) do { if (ipu_log_level) printf(TAG fmt, ##args); } while (0)